go.sum

```diff
@@ -557,17 +538,12 @@ github.com/hashicorp/go-multierror v1.1.1/go.mod h1:iw975J/qwKPdAO1clOe2L8331t/9
 github.com/hashicorp/go-retryablehttp v0.7.0/go.mod h1:vAew36LZh98gCBJNLH42IQ1ER/9wtLZZ8meHqQvEYWY=
 github.com/hashicorp/go-retryablehttp v0.7.7 h1:C8hUCYzor8PIfXHa4UrZkU4VvK8o9ISHxT2Q8+VepXU=
 github.com/hashicorp/go-retryablehttp v0.7.7/go.mod h1:pkQpWZeYWskR+D1tR2O5OcBFOxfA7DoAO6xtkuQnHTk=
-github.com/hashicorp/go-version v1.5.0/go.mod h1:fltr4n8CU8Ke44wwGCBoEymUuxUHl09ZGVZPK5anwXA=
 github.com/hashicorp/go-version v1.6.0 h1:feTTfFNnjP967rlCxM/I9g701jU+RN74YKx2mOkIeek=
 github.com/hashicorp/go-version v1.6.0/go.mod h1:fltr4n8CU8Ke44wwGCBoEymUuxUHl09ZGVZPK5anwXA=
 github.com/hashicorp/golang-lru v0.5.0/go.mod h1:/m3WP610KZHVQ1SGc6re/UDhFvYD7pJ4Ao+sR/qLZy8=
 github.com/hashicorp/golang-lru v0.5.1/go.mod h1:/m3WP610KZHVQ1SGc6re/UDhFvYD7pJ4Ao+sR/qLZy8=
-github.com/hashicorp/hc-install v0.4.0 h1:cZkRFr1WVa0Ty6x5fTvL1TuO1flul231rWkGH92oYYk=
-github.com/hashicorp/hc-install v0.4.0/go.mod h1:5d155H8EC5ewegao9A4PUTMNPZaq+TbOzkJJZ4vrXeI=
 github.com/hashicorp/terraform-exec v0.17.3 h1:MX14Kvnka/oWGmIkyuyvL6POx25ZmKrjlaclkx3eErU=
 github.com/hashicorp/terraform-exec v0.17.3/go.mod h1:+NELG0EqQekJzhvikkeQsOAZpsw0cv/03rbeQJqscAI=
-github.com/hashicorp/terraform-json v0.14.0 h1:sh9iZ1Y8IFJLx+xQiKHGud6/TSUCM0N8e17dKDpqV7s=
-github.com/hashicorp/terraform-json v0.14.0/go.mod h1:5A9HIWPkk4e5aeeXIBbkcOvaZbIYnAIkEyqP2pNSckM=
```

Comment:
I haven't dug around to see why these are getting dropped, but I don't see how they'd be related to my openshift/api bump.  Maybe an earlier pull dropped their imports, and nobody happened to run `go mod tidy` and `go mod vendor` until now?

Reply:
[images presubmit][1]:

```
vendor/github.com/hashicorp/terraform-exec/tfexec/providers_schema.go:7:2: cannot find module providing package github.com/hashicorp/terraform-json: import lookup disabled by -mod=vendor
```

So I'm misunderstanding how installer vendor bumps are supposed to happen...

[1]: https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_installer/9002/pull-ci-openshift-installer-master-images/1834316438404337664#1:build-log.txt%3A1827

Reply:
Hmm, but I also seem to be following [the doc'ed process][1] (with an additional `go generate ./pkg/types/installconfig.go` that I can add to the docs in follow-up, `NO-ISSUE` work).  Not clear to me what I'm missing...

[1]: https://github.com/openshift/installer/blob/939ea5c171f121fa6d1e5d5b2b1a5f0656506691/docs/dev/dependencies.md

Reply:
Poking around:

* [`github.com/hashicorp/terraform-exec/tfexec` imports `github.com/hashicorp/terraform-json`][2].
* [`pkg/terraform` imports `github.com/hashicorp/terraform-exec/tfexec`][1].
* [`pkg/infrastructure/platform` imports `pkg/terraform`][0]. 
* [`pkg/asset/cluster` imports `pkg/infrastructure/platform`][a].
* [`pkg/asset/targets` imports `pkg/asset/cluster`][b].
* Only [`pkg/asset/store/assetcreate_test.go` imports `pkg/asset/targets`][c].

Hmm, so maybe there's a knob for "also vendor the things I need for `*_test.go`?

[c]: https://github.com/openshift/installer/blob/939ea5c171f121fa6d1e5d5b2b1a5f0656506691/pkg/asset/store/assetcreate_test.go#L13
[b]: https://github.com/openshift/installer/blob/939ea5c171f121fa6d1e5d5b2b1a5f0656506691/pkg/asset/targets/targets.go#L5
[a]: https://github.com/openshift/installer/blob/939ea5c171f121fa6d1e5d5b2b1a5f0656506691/pkg/asset/cluster/cluster.go#L29
[0]: https://github.com/openshift/installer/blob/939ea5c171f121fa6d1e5d5b2b1a5f0656506691/pkg/infrastructure/platform/platform.go#L20
[1]: https://github.com/openshift/installer/blob/939ea5c171f121fa6d1e5d5b2b1a5f0656506691/pkg/terraform/terraform.go#L10
[2]: https://github.com/openshift/installer/blob/939ea5c171f121fa6d1e5d5b2b1a5f0656506691/vendor/github.com/hashicorp/terraform-exec/tfexec/validate.go#L9

Reply:
Walking docs, [`go mod vendor`][1]:

> The go mod vendor command constructs a directory named vendor in the [main module’s](https://go.dev/ref/mod#glos-main-module) root directory that contains copies of all packages needed to support builds and tests of packages in the main module.

[main module][2]:

> main module: The module in which the go command is invoked. The main module is defined by a [go.mod file](https://go.dev/ref/mod#glos-go-mod-file) in the current directory or a parent directory. See [Modules, packages, and versions](https://go.dev/ref/mod#modules-overview).

[`go.mod` file][3]:

> `go.mod` file: The file that defines a module’s path, requirements, and other metadata. Appears in the [module’s root directory](https://go.dev/ref/mod#glos-module-root-directory). See the section on [`go.mod` files](https://go.dev/ref/mod#go-mod-file).

Not helping me yet.  Moving over to [`go mod tidy`][4]:

> `go mod tidy` ensures that the `go.mod` file matches the source code in the module. It adds any missing module requirements necessary to build the current module’s packages and dependencies, and it removes requirements on modules that don’t provide any relevant packages. It also adds any missing entries to `go.sum` and removes unnecessary entries.
> ...
> `go mod tidy` works by loading all of the packages in the [main module](https://go.dev/ref/mod#glos-main-module) and all of the packages they import, recursively. This includes packages imported by tests (including tests in other modules). `go mod tidy` acts as if all build tags are enabled, so it will consider platform-specific source files and files that require custom build tags, even if those source files wouldn’t normally be built. There is one exception: the `ignore` build tag is not enabled, so a file with the build constraint `// +build ignore` will not be considered. Note that `go mod tidy` will not consider packages in the main module in directories named testdata or with names that start with `.` or `_` unless those packages are explicitly imported by other packages.

But still not clear to me what I'm holding wrong...

[1]: https://go.dev/ref/mod#go-mod-vendor
[2]: https://go.dev/ref/mod#glos-main-module
[3]: https://go.dev/ref/mod#glos-go-mod-file
[4]: https://go.dev/ref/mod#go-mod-tidy

Reply:
And dropping back to the main branch:

```console
$ go version
go version go1.22.4 linux/amd64
$ git --no-pager log -1 --date=short --format='%ad %h %s'
2024-09-11 929f21c3dc Merge pull request #8830 from shiftstack/openstack-install-from-repo
$ go mod why github.com/hashicorp/terraform-json
# github.com/hashicorp/terraform-json
(main module does not need package github.com/hashicorp/terraform-json)
```

So... I guess I can try and bisect that?

Reply:
Apparently it was a Go issue, because I bumped from Go 1.22.4 to 1.23.1, and no longer get these weird deletions.